forward/rewriting.lua

```diff
@@ -103,6 +74,84 @@ function Rewriting:push()
    end
 end
 
+-- Parses a datagram to compute the packet's five tuple and the backend
+-- pool to forward to packet to. Due to IP fragmentation, the packet to
+-- be forwarded may be different from the packet received.
+-- See reassembly.lua for an explanation of IP fragmentation handling.
+-- Expected input datagram structure:
+-- IPv4 fragment --
+--    Ethernet (popped) | IPv4 (parsed) | payload
+-- Redirected IPv4 fragment --
+--    Ethernet (popped) | IPv4 (parsed) | GRE | IPv4 | payload
+-- Full packet --
+--    Ethernet (popped) | IPv4/IPv6 (parsed) | TCP/UDP | payload
+-- Expected output datagram structure:
+--    Ethernet (popped/missing) | IPv4 | TCP/UDP (parsed) | payload
+-- Arguments:
+-- datagram (datagram) -- Datagram to process. Should have Ethernet
+--    header popped and IP header parsed.
+-- ip_header (header) -- Parsed IP header of datagram.
+-- ip_type (int) -- IP header type, can be L3_IPV4 or L3_IPV6
+-- Returns:
+-- forward_datagram (bool) -- Indicates whether there is a datagram to
+--    forward.
+-- t (binary) -- Five tuple of datagram.
+-- t_len (int) -- Length of t.
+-- backend_pool (int) -- Backend pool to forward the packet to.
+-- new_datagram (datagram) -- Datagram to forward. Should have Ethernet
+--    header popped or missing, and TCP/UDP headers parsed.
+-- new_datagram_len (int) -- Length of new_datagram.
+function Rewriting:handle_fragmentation_and_get_forwarding_params(datagram, ip_header, ip_type)
+   local ip_src = ip_header:src()
+   local ip_dst = ip_header:dst()
+   local l4_type
+   if ip_type == L3_IPV4 then
+      l4_type = ip_header:protocol()
+   else
+      l4_type = ip_header:next_header()
+   end
+   local ip_total_length = ip_header:total_length()
+   local prot_class = ip_header:upper_layer()
+
+   local frag_off = ip_header:frag_off()
+   local mf = band(ip_header:flags(), IP_MF_FLAG) ~= 0
+   if frag_off ~= 0 or mf then
+      -- Packet is an IPv4 fragment; redirect to another spike
+      -- Set ports to zero to get three-tuple
+      local t3, t3_len = five_tuple(ip_type, ip_src, 0, ip_dst, 0)
+      -- TODO: Return spike backend pool
+      return true, t3, t3_len, nil, datagram, ip_total_length
```

Comment:
This does not conform to the stated output structure because there's no parsed L4 header.
In fact it would be impossible in this case since no L4 header even exists to parse.
Probably we want to treat L4 as the payload rather than returning it parsed.